### PR TITLE
[8.x] Optimize PHP performance by using fully-qualified function calls

### PR DESCRIPTION
**Pull request by claudiodekker**
This PR enables StyleCI's [`native_function_invocation_symfony`](https://docs.styleci.io/fixers#native_function_invocation_symfony) fixer & introduces changes to the way native PHP functions are referenced.

By directly referencing these functions in their (global) namespace (`\is_null`) instead of calling them within the current scope (`is_null`), PHP can skip the step of resolving this function within the current namespace. More importantly, it also allows PHP's OPCache to cache these calls, which leads to significant performance improvements.

We've benchmarked these changes on a [Forge](https://forge.laravel.com) provisioned server on DigitalOcean, running a similar setup [as described here](https://medium.com/@taylorotwell/benchmarking-laravel-symfony-zend-2c01c2b270f8). Just to recap, this means our configuration is looking like this:

- PHP 7.4, with OPCache enabled.
- `web` (session) middleware (`app/Http/Kernel.php`) disabled.
- Optimized composer autoloader (`composer dumpautoload -o`)
- Cached configuration (`php artisan config:cache`)

Furthermore, to benchmark, we're running `ab -t 10 -c 10 http://server.address/` against the homepage (`/`) of an otherwise fresh [`laravel/laravel`](https://github.com/laravel/laravel) installation.

To make sure that we weren't catching the server in a bad moment, we ran the benchmark five times each at least a minute apart, and gathered the **mean** requests per second of every run:

| # | Before | After |
| -- | -- | -- |
| 1 | 801,77 | 937,17 |
| 2 | 717,24 | 681,92 |
| 3 | 587,44 | 689,93 |
| 4 | 554,62 | 674,22 |
| 5 | 469,30 | 947,28 |
| 🔥  | **626,074** | **786,104** |

That's a **25.56%** performance improvement on a fresh Laravel v8.44.0
When introducing a database to the mix, the gains drop dramatically to only 2,63% (see https://github.com/laravel/framework/pull/37517#issuecomment-853062134)

(Inspired by https://github.com/scssphp/scssphp/issues/96)

### CODE REVIEW

**Comment by eusonlito** [comment]
Only as comment, on my deploys I have a `php-cs-fixer` command to fix all `vendor` folder with `native_constant_invocation` and `native_function_invocation`:

### .php-cs-fixer.php

```php
<?php declare(strict_types=1);

use PhpCsFixer\Finder;
use PhpCsFixer\Config;

return (new Config())
    ->setRiskyAllowed(true)
    ->setRules(['native_constant_invocation' => true, 'native_function_invocation' => true])
    ->setFinder(Finder::create()->exclude(['phpunit', 'php-parallel-lint', 'tests'])->in('vendor'));
```

### Execution inside deploy script:

```
$> php ./vendor/bin/php-cs-fixer fix --ansi --quiet
```

The performance difference is very noticeable.

**Comment by GrahamCampbell** [comment]
@eusonlito That is a very risky thing to do, and will break octan, for example.

File: src/Illuminate/Cache/PhpRedisLock.php [hunk 1]
```diff
@@ -51,11 +51,11 @@ protected function serializedAndCompressedOwner(): string
         // https://github.com/phpredis/phpredis/issues/1938
         if ($this->compressed()) {
             if ($this->lzfCompressed()) {
-                $owner = \lzf_compress($owner);
+                $owner = lzf_compress($owner);
```
Comment (s-patompong) [inline src/Illuminate/Cache/PhpRedisLock.php:54]: Just wondering, why would it remove the \ from this line if the purpose of this fixer rule is to 'Add leading slash before optimized function invocation to speed up resolving.' according to their doc.

Not a reviewer or anything, just wonder :)

**Comment by s-patompong** [comment]
> @eusonlito That is a very risky thing to do, and will break octan, for example.

Yeah, the doc also mentioned this:
```
This is a "risky" heuristic fixer, and could change code behavior!
```

**Comment by eusonlito** [comment]
> @eusonlito That is a very risky thing to do, and will break octan, for example.

@GrahamCampbell thanks for your advice. I haven't tried Octane yet, but in about 5 years using this method in about 50 or 60 Laravel projects, never notice any issue.

Anyway, thanks. I will check it.

**Comment by GrahamCampbell** [comment]
It specifically breaks the ability to override core functions. Uploading files over HTTP in octane will break if you run that command on the vendor directory.

**Comment by eusonlito** [comment]
> It specifically breaks the ability to override core functions. Uploading files over HTTP in octane will break if you run that command on the vendor directory.

If my memory doesn't fail, I think that Guzzle had same *problem* on version 6 and was solved on code at version 7.

**Comment by deleugpn** [comment]
Is it a boost enough to justify the ugly codebase though? 🤷‍♂️

**Comment by foremtehan** [comment]
Can anyone explain why it cant be done in php core  itself ?

**Comment by claudiodekker** [comment]
> But is it a boost enough to justify the ugly codebase though? 🤷‍♂️

Yeah, that's indeed the question.. Internally, I already mentioned that it'd be great if there was a way to have these be auto-imported when referenced (`use \is_null;`), as then it'd only occur a few time near the top of the file. In theory, that should also be just-as-quick.

> Can anyone explain why it cant be done in php core itself ? I mean can't they add performance by default ?

They probably can, this is just one remnants from an older PHP version. It's kinda silly (IMO) to re-define built-in langauge functions in your local namespace anyway ¯\_(ツ)_/¯

File: src/Illuminate/Cache/PhpRedisLock.php (hunk 1 above)
Comment (crynobone) [reply src/Illuminate/Cache/PhpRedisLock.php:54]: because the config will only prefix `\` for global functions that can be improved via opcache.

**Comment by timacdonald** [comment]
> Yeah, that's indeed the question.. Internally, I already mentioned that it'd be great if there was a way to have these be auto-imported when referenced (use \is_null;), as then it'd only occur a few time near the top of the file. In theory, that should also be just-as-quick.

Totally possible and something I do on my projects instead of messing with the app code itself.

See: https://github.com/timacdonald/log-fake/blob/master/src/LogFake.php#L11

Via the PHP-CS-Fixer rule: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/3.0/doc/rules/import/global_namespace_import.rst

I'd really prefer to see that implementation personally and keep the mess at the top of the file rather than scattered throughout.

File: src/Illuminate/Cache/PhpRedisLock.php (hunk 1 above)
Comment (s-patompong) [reply src/Illuminate/Cache/PhpRedisLock.php:54]: I see, thanks for the clarification!

**Comment by taylorotwell** [comment]
Do we still see any measurable performance benefit when adding in for example 1 Eloquent DB call into the request lifecycle?

**Comment by Krisell** [comment]
> Do we still see any measurable performance benefit when adding in for example 1 Eloquent DB call into the request lifecycle?

Waiting for DB-response doesn't use CPU-time so even if the difference is much smaller, or even insignificant, that doesn't mean the cpu load decrease is also insignificant. But it would definitely be interesting to see the numbers.

File: src/Illuminate/Cache/PhpRedisLock.php (hunk 1 above)
Comment (Krisell) [reply src/Illuminate/Cache/PhpRedisLock.php:54]: Considering that `lzf_compress` still has to be resolved via the namespace, is the result of that process cached in opcache (resulting in the same effect as prepending with \\), or is the performance effect of these calls just insignificant when compared to the functions that have special opcodes?

**Comment by garygreen** [comment]
> [in reference to having this be built into core PHP] That said, IMO it's still kinda silly to even want the ability to override built-in language functions within your own namespaces, but there's probably a real use-case that I'm not aware of out there somewhere ¯_(ツ)_/¯

I totally agree it does seem like every PHP codebase is being slowed down for that 0.01% edge case who override built in php functions based on namespaces.

Rather than importing for each file as you suggested `use \is_null` etc instead just cater to those edge cases who want to override php builtins `declare(override_builtins = 1);` - suggest default value should be `0`. That way `\is_null` etc is not needed by default for any PHP class.

Would love to see an RFC.

**Comment by taylorotwell** [comment]
If someone could do some benchmarks on an application that makes 1 database query please let me know.

**Comment by eusonlito** [comment]
> If someone could do some benchmarks on an application that makes 1 database query please let me know.

I have created some "advanced" tests: https://gist.github.com/eusonlito/961e6e0a8dde3736fca5465be6fc2c64

**Comment by taylorotwell** [comment]
@eusonlito sorry - a lot of information there.

I am primarily interested in what is the total execution time of handling an HTTP request before and after this change when hitting a route with 1 database query.

**Comment by eusonlito** [comment]
@taylorotwell.

Here:

```
before - http://localhost/laravel/public/test/query/1

Requests per second:    882.14 [#/sec] (mean)
Time per request:       11.336 [ms] (mean)
Time per request:       1.134 [ms] (mean, across all concurrent requests)
```
```
after - http://localhost/laravel/public/test/query/1

Requests per second:    901.76 [#/sec] (mean)
Time per request:       11.089 [ms] (mean)
Time per request:       1.109 [ms] (mean, across all concurrent requests)

**Comment by garygreen** [comment]
That's only a 2% speed difference which puts it in the region of a micro optimization?

I am kind of surprised seeing @claudiodekker result of 25% speed difference. I'm highly sceptical this has skipped the attention of PHP maintainers, as such a speed increase would be an easy win for most PHP projects if this was configurable in some way.

Is it a possibility the original results are down to testing method, bad timing when running tests, other factors at play?

**Comment by claudiodekker** [comment]
No, the original results are accurate. I'm about to run another test with a single DB query against both. 
Will put both instances up so people can verify the results themselves.

**Comment by claudiodekker** [comment]
Updated results, using the DB branches I referred to in my previous comment:

| # | Before | After |
| -- | -- | -- |
| 1 | 353,49 | 364,29 |
| 2 | 363,89 | 360,46 |
| 3 | 356,39 | 374,08 |
| 4 | 353,76 | 358,54 |
| 5 | 357,18 | 374,35 |
| 🔥  | **356,942** | **366,344** |

As @garygreen & @eusonlito's results indeed already conclude, the moment a database is introduced, the gains drops all the way down to only **2,63%**.

**Comment by taylorotwell** [comment]
@eusonlito @claudiodekker 

Was this with or without opcache enabled?

**Comment by claudiodekker** [comment]
@taylorotwell With OPCache enabled, as are all the tests above. Same exact conditions as the original post.

**Comment by eusonlito** [comment]
@taylorotwell previous tests are in an real life project.

Here tests with a clean installation Laravel and only adding a query to default route (OPCache enabled):

```php
Route::get('/index', function () {
    DB::table('exchange')->limit(1)->get();
    return view('welcome');
});
```

```
# PHP 7.4.18 - Before

Requests per second:    1177.26 [#/sec] (mean)
Time per request:       8.494 [ms] (mean)
Time per request:       0.849 [ms] (mean, across all concurrent requests)
```
```
# PHP 7.4.18 - After

Requests per second:    1190.68 [#/sec] (mean)
Time per request:       8.399 [ms] (mean)
Time per request:       0.840 [ms] (mean, across all concurrent requests)
```

```
# PHP 8.0.5 - Before

Requests per second:    1184.53 [#/sec] (mean)
Time per request:       8.442 [ms] (mean)
Time per request:       0.844 [ms] (mean, across all concurrent requests)
```
```
# PHP 8.0.5 - After

Requests per second:    1211.61 [#/sec] (mean)
Time per request:       8.253 [ms] (mean)
Time per request:       0.825 [ms] (mean, across all concurrent requests)
```
Without query:
```php
Route::get('/index', function () {
    return view('welcome');
});
```
```
# PHP 7.4.18 - Before

Requests per second:    1522.38 [#/sec] (mean)
Time per request:       6.569 [ms] (mean)
Time per request:       0.657 [ms] (mean, across all concurrent requests)
```
```
# PHP 7.4.18 - After

Requests per second:    1549.06 [#/sec] (mean)
Time per request:       6.456 [ms] (mean)
Time per request:       0.646 [ms] (mean, across all concurrent requests)
```

```
# PHP 8.0.5 - Before

Requests per second:    1508.91 [#/sec] (mean)
Time per request:       6.627 [ms] (mean)
Time per request:       0.663 [ms] (mean, across all concurrent requests)
```
```
# PHP 8.0.5 - After

Requests per second:    1535.17 [#/sec] (mean)
Time per request:       6.514 [ms] (mean)
Time per request:       0.651 [ms] (mean, across all concurrent requests)
```
And without view render:
```php
Route::get('/index', function () {
    DB::table('exchange')->limit(1)->get();
});
```
```
# PHP 7.4.18 - Before

Requests per second:    1252.13 [#/sec] (mean)
Time per request:       7.986 [ms] (mean)
Time per request:       0.799 [ms] (mean, across all concurrent requests)
```
```
# PHP 7.4.18 - After

Requests per second:    1293.12 [#/sec] (mean)
Time per request:       7.733 [ms] (mean)
Time per request:       0.773 [ms] (mean, across all concurrent requests)
```

```
# PHP 8.0.5 - Before

Requests per second:    1257.99 [#/sec] (mean)
Time per request:       7.949 [ms] (mean)
Time per request:       0.795 [ms] (mean, across all concurrent requests)
```
```
# PHP 8.0.5 - After

Requests per second:    1296.82 [#/sec] (mean)
Time per request:       7.711 [ms] (mean)
Time per request:       0.771 [ms] (mean, across all concurrent requests)
```

**Comment by allowing** [comment]
Seeing this discussion, I also wrote a script to try to optimize it. I optimized the entire `vendor`. But the subjective feeling did not get faster. 😂

```php
<?php

\ini_set('memory_limit', -1);

$f = function ($dir, array &$user_funcs) use (&$f) {
    $h = \opendir($dir);

    while (false !== $file = \readdir($h)) {
        if (\in_array($file, ['.', '..'])) {
            continue;
        }

        if (\is_dir($filename = "$dir/$file")) {
            $f($filename, $user_funcs);
            continue;
        }

        if (\strtolower(\substr($file, -4)) != '.php') {
            continue;
        }

        $content = \file_get_contents($filename);

        if (!preg_match('/namespace\s+\w+/i', $content)) {
            continue;
        }

        \preg_match_all('/\w*\s*function\s+(\w+)\s*\(/i', $content, $matches);

        foreach ($matches[0] as $k => $match) {
            if (\preg_match('/(public|private|protected|static)/i', $match)) {
                continue;
            }
            $user_funcs[] = $matches[1][$k];
        }
    }
};

$user_funcs = [];
$f(__DIR__ . '/vendor', $user_funcs);
$user_funcs = \array_unique($user_funcs);

// \sort($user_funcs);
// \file_put_contents(__DIR__ . '/user_funcs.txt', \implode("\n", $user_funcs));die;

if (!$user_funcs) {
    die('$user_funcs is empty!');
}

$finder = (new class extends \Symfony\Component\Finder\Finder {
    public function __construct()
    {
        parent::__construct();

        $this->files()->name('*.php');
    }
})->in(__DIR__ . '/vendor');

return (new \PhpCsFixer\Config())
    ->setRules([
        'native_function_invocation' => [
            'exclude' => $user_funcs,
            'include' => ['@all'],
            'scope' => 'all',
            // Whether leading \ of function call not meant to have it should be removed.
            'strict' => false,
        ],
    ])
    ->setRiskyAllowed(true)
    ->setUsingCache(false)
    ->setFinder($finder);
```

**Comment by taylorotwell** [comment]
I think if we are seeing pretty minor improvements once we introduce a single query - we may hold off on this for now. Those improvements will become even less noticeable once multiple queries are introduced, calls to the queue system, third party API calls, etc.

**Comment by Krisell** [comment]
> I think if we are seeing pretty minor improvements once we introduce a single query - we may hold off on this for now. Those improvements will become even less noticeable once multiple queries are introduced, calls to the queue system, third party API calls, etc.

That depends on how "noticeable" is defined, of course not for a user (which is what response time measures), but certainly for the server given the initial measurements here and since none of these remote calls consume CPU cycles while waiting for response. I don't know of any better way to make measurements though (except for only running CPU-tests, i.e. without DB/networking calls, like some tests here).

To clarify my point, consider a 1 second DB-call being included in the benchmarking data – then an update from php 5 to 7 would also not be noticeable.

**Pull request closed without merging**